// @freesewing/components/SampleConfigurator/index.js
/**
 * @freesewing/components | v2.19.5
 * A collection of React components for FreeSewing web UIs
 * (c) 2021 Joost De Cock <joost@joost.at> (https://github.com/joostdecock)
 * @license MIT
 */
'use strict';

var React = require('react');
var reactIntl = require('react-intl');
var models = require('@freesewing/models');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function _typeof(obj) {
  "@babel/helpers - typeof";

  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function (obj) {
      return typeof obj;
    };
  } else {
    _typeof = function (obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    };
  }

  return _typeof(obj);
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];

  if (_i == null) return;
  var _arr = [];
  var _n = true;
  var _d = false;

  var _s, _e;

  try {
    for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

function _createForOfIteratorHelper(o, allowArrayLike) {
  var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"];

  if (!it) {
    if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
      if (it) o = it;
      var i = 0;

      var F = function () {};

      return {
        s: F,
        n: function () {
          if (i >= o.length) return {
            done: true
          };
          return {
            done: false,
            value: o[i++]
          };
        },
        e: function (e) {
          throw e;
        },
        f: F
      };
    }

    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }

  var normalCompletion = true,
      didErr = false,
      err;
  return {
    s: function () {
      it = it.call(o);
    },
    n: function () {
      var step = it.next();
      normalCompletion = step.done;
      return step;
    },
    e: function (e) {
      didErr = true;
      err = e;
    },
    f: function () {
      try {
        if (!normalCompletion && it.return != null) it.return();
      } finally {
        if (didErr) throw err;
      }
    }
  };
}

var OptionGroup = function OptionGroup(props) {
  return /*#__PURE__*/React__default['default'].createElement(React__default['default'].Fragment, null, props.options.map(function (name) {
    var output = [];

    if (_typeof(name) === 'object') {
      // Subgroup
      for (var _i = 0, _Object$keys = Object.keys(name); _i < _Object$keys.length; _i++) {
        var subGroup = _Object$keys[_i];
        output.push( /*#__PURE__*/React__default['default'].createElement("h5", {
          key: subGroup + '-title'
        }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
          id: 'optiongroups.' + subGroup
        })));
        var children = [];

        var _iterator = _createForOfIteratorHelper(name[subGroup]),
            _step;

        try {
          var _loop = function _loop() {
            var option = _step.value;
            children.push( /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
              href: "#logo",
              onClick: function onClick() {
                return props.sampleOption(option);
              },
              className: props.activeOption === option ? 'active' : ''
            }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
              id: 'options.' + props.config.name + '.' + option + '.title'
            }))));
          };

          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            _loop();
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }

        output.push( /*#__PURE__*/React__default['default'].createElement("ul", {
          style: {
            paddingLeft: '1rem'
          }
        }, children));
      }
    } else output.push( /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
      href: "#logo",
      onClick: function onClick() {
        return props.sampleOption(name);
      },
      className: props.activeOption === name ? 'active' : ''
    }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: 'options.' + props.config.name + '.' + name + '.title'
    }))));

    return output;
  }));
};

var OptionGroup$1 = reactIntl.injectIntl(OptionGroup);

var PatternOptions = function PatternOptions(props) {
  var renderGroup = function renderGroup(group) {
    var output = [];
    var children = /*#__PURE__*/React__default['default'].createElement("ul", null, /*#__PURE__*/React__default['default'].createElement(OptionGroup$1, {
      key: group + '-group',
      units: props.units,
      config: props.config,
      options: props.config.optionGroups[group],
      sampleOption: props.sampleOption,
      activeOption: props.activeOption
    }));
    output.push( /*#__PURE__*/React__default['default'].createElement("li", {
      key: group + '-ghead'
    }, /*#__PURE__*/React__default['default'].createElement("h6", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: 'optiongroups.' + group
    })), children));
    return output;
  };

  return /*#__PURE__*/React__default['default'].createElement("ul", null, Object.keys(props.config.optionGroups).map(function (group) {
    return renderGroup(group);
  }));
};

var nonHuman = {
  withoutBreasts: {
    dolls: {},
    giants: {}
  },
  withBreasts: {
    dolls: {},
    giants: {}
  }
};

var round = function round(val) {
  return Math.round(val * 10) / 10;
};

for (var i = 0.1; i < 1; i += 0.1) {
  var name = "Doll ".concat(Math.round(i * 10), "/10");
  nonHuman.withBreasts.dolls[name] = {}; // withBreasts: Based on Anneke (size 34)

  for (var _i = 0, _Object$entries = Object.entries(models.withBreasts.size34); _i < _Object$entries.length; _i++) {
    var _Object$entries$_i = _slicedToArray(_Object$entries[_i], 2),
        m = _Object$entries$_i[0],
        val = _Object$entries$_i[1];

    nonHuman.withBreasts.dolls[name][m] = m === 'shoulderSlope' ? val : round(val * i);
  } // withoutBreasts: Based on Ronan (size 42)


  nonHuman.withoutBreasts.dolls[name] = {};

  for (var _i2 = 0, _Object$entries2 = Object.entries(models.withoutBreasts.size42); _i2 < _Object$entries2.length; _i2++) {
    var _Object$entries2$_i = _slicedToArray(_Object$entries2[_i2], 2),
        _m = _Object$entries2$_i[0],
        _val = _Object$entries2$_i[1];

    nonHuman.withoutBreasts.dolls[name][_m] = _m === 'shoulderSlope' ? _val : round(_val * i);
  }
}

for (var _i3 = 1; _i3 <= 2.5; _i3 += 0.5) {
  var _name = "Giant ".concat(_i3, "/1");

  nonHuman.withBreasts.giants[_name] = {}; // withBreasts: Based on Anneke (size 34)

  for (var _i4 = 0, _Object$entries3 = Object.entries(models.withBreasts.size34); _i4 < _Object$entries3.length; _i4++) {
    var _Object$entries3$_i = _slicedToArray(_Object$entries3[_i4], 2),
        _m2 = _Object$entries3$_i[0],
        _val2 = _Object$entries3$_i[1];

    nonHuman.withBreasts.giants[_name][_m2] = _m2 === 'shoulderSlope' ? _val2 : round(_val2 * _i3);
  }

  nonHuman.withoutBreasts.giants[_name] = {}; // withoutBreasts: Based on Ronan (size 42)

  for (var _i5 = 0, _Object$entries4 = Object.entries(models.withoutBreasts.size42); _i5 < _Object$entries4.length; _i5++) {
    var _Object$entries4$_i = _slicedToArray(_Object$entries4[_i5], 2),
        _m3 = _Object$entries4$_i[0],
        _val3 = _Object$entries4$_i[1];

    nonHuman.withoutBreasts.giants[_name][_m3] = _m3 === 'shoulderSlope' ? _val3 : round(_val3 * _i3);
  }
}

var SampleConfigurator = function SampleConfigurator(props) {
  var _useState = React.useState(),
      _useState2 = _slicedToArray(_useState, 2),
      type = _useState2[0],
      setType = _useState2[1];

  var _useState3 = React.useState(),
      _useState4 = _slicedToArray(_useState3, 2),
      instance = _useState4[0],
      setInstance = _useState4[1];

  var sampleOption = function sampleOption(option) {
    setType('option');
    setInstance(option);
    props.updateGist({
      type: 'option',
      option: option
    }, 'settings', 'sample');
  };

  var sampleMeasurement = function sampleMeasurement(measurement) {
    setType('measurement');
    setInstance(measurement);
    props.updateGist({
      type: 'measurement',
      measurement: measurement
    }, 'settings', 'sample');
  };

  var sampleModels = function sampleModels(models) {
    setType('models');
    props.updateGist({
      type: 'models',
      models: models
    }, 'settings', 'sample');
  };

  return /*#__PURE__*/React__default['default'].createElement("ul", null, /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("h5", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.patternOptions"
  })), /*#__PURE__*/React__default['default'].createElement(PatternOptions, {
    config: props.config,
    gist: props.gist,
    sampleOption: sampleOption,
    activeOption: type === 'option' ? instance : null
  })), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("h5", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.measurements"
  })), /*#__PURE__*/React__default['default'].createElement("ul", null, props.config.measurements.map(function (m) {
    return /*#__PURE__*/React__default['default'].createElement("li", {
      key: m
    }, /*#__PURE__*/React__default['default'].createElement("a", {
      href: "#logo",
      onClick: function onClick() {
        return sampleMeasurement(m);
      },
      className: type === 'measurement' && instance === m ? 'active' : ''
    }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: 'measurements.' + m
    })));
  }))), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("h5", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.people"
  })), /*#__PURE__*/React__default['default'].createElement("ul", null, /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
    href: "#logo",
    onClick: function onClick() {
      sampleModels(models.withBreasts);
      setType('models');
      setInstance('withBreasts');
    },
    className: type === 'models' && instance === 'withBreasts' ? 'active' : ''
  }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.withBreasts"
  }))), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
    href: "#logo",
    onClick: function onClick() {
      sampleModels(models.withoutBreasts);
      setType('models');
      setInstance('withoutBreasts');
    },
    className: type === 'models' && instance === 'withoutBreasts' ? 'active' : ''
  }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.withoutBreasts"
  }))))), ['dolls', 'giants'].map(function (type) {
    return /*#__PURE__*/React__default['default'].createElement("li", {
      key: type
    }, /*#__PURE__*/React__default['default'].createElement("h5", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: "app.".concat(type)
    })), /*#__PURE__*/React__default['default'].createElement("ul", null, /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
      href: "#logo",
      onClick: function onClick() {
        sampleModels(nonHuman.withBreasts[type]);
        setType('models');
        setInstance('withBreasts');
      },
      className: type === 'models' && instance === 'withBreasts' ? 'active' : ''
    }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: "app.withBreasts"
    }))), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
      href: "#logo",
      onClick: function onClick() {
        sampleModels(nonHuman.withoutBreasts[type]);
        setType('models');
        setInstance('withoutBreasts');
      },
      className: type === 'models' && instance === 'withoutBreasts' ? 'active' : ''
    }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: "app.withoutBreasts"
    })))));
  }));
};

module.exports = SampleConfigurator;
